lib/event-libs/sdevent/sdevent.c

```diff
@@ -0,0 +1,400 @@
+};
+
+struct lws_wsi_watcher_sdevent {
+    struct sd_event_source *source;
+    uint32_t events;
+};
+
+static int sultimer_handler(sd_event_source *s, uint64_t usec, void *userdata) {
+    struct lws_context_per_thread *pt = (struct lws_context_per_thread*) userdata;
+
+    lws_usec_t us;
+
+    lws_context_lock(pt->context, __func__);
+    lws_pt_lock(pt, __func__);
+    us = __lws_sul_service_ripe(pt->pt_sul_owner, LWS_COUNT_PT_SUL_OWNERS,
+                                lws_now_usecs());
+    if (us) {
+        uint64_t alarmTime;
+        sd_event_now(sd_event_source_get_event(s), CLOCK_MONOTONIC, &alarmTime);
+        alarmTime += us;
+        sd_event_source_set_time(pt_to_priv_sd(pt)->sultimer, alarmTime);
+        sd_event_source_set_enabled(pt_to_priv_sd(pt)->sultimer, SD_EVENT_ONESHOT);
+    }
+
+    lws_pt_unlock(pt);
+    lws_context_unlock(pt->context);
+
+    return 0;
+}
+
+static int idle_handler(sd_event_source *s, uint64_t usec, void *userdata) {
+    struct lws_context_per_thread *pt = (struct lws_context_per_thread*) userdata;
+
+    lws_usec_t us;
+
+    lws_service_do_ripe_rxflow(pt);
+
+    lws_context_lock(pt->context, __func__);
+    lws_pt_lock(pt, __func__);
+
+    /*
+     * is there anybody with pending stuff that needs service forcing?
+     */
+    if (!lws_service_adjust_timeout(pt->context, 1, pt->tid)) {
+        /* -1 timeout means just do forced service */
+        _lws_plat_service_forced_tsi(pt->context, pt->tid);
+    }
+
+    /* account for sultimer */
+
+    us = __lws_sul_service_ripe(pt->pt_sul_owner, LWS_COUNT_PT_SUL_OWNERS,
+                                lws_now_usecs());
+
+    if (us) {
+        uint64_t alarmTime;
+        sd_event_now(sd_event_source_get_event(s), CLOCK_MONOTONIC, &alarmTime);
+        alarmTime += us;
+        sd_event_source_set_time(pt_to_priv_sd(pt)->sultimer, alarmTime);
+        sd_event_source_set_enabled(pt_to_priv_sd(pt)->sultimer, SD_EVENT_ONESHOT);
+    }
+
+    sd_event_source_set_enabled(pt_to_priv_sd(pt)->idletimer, SD_EVENT_OFF);
+
+    lws_pt_unlock(pt);
+    lws_context_unlock(pt->context);
+
+    return 0;
+}
+
+static int sock_accept_handler(sd_event_source *s, int fd, uint32_t revents, void *userdata) {
+    struct lws *wsi = (struct lws*) userdata;
+    struct lws_context *context = wsi->a.context;
+    struct lws_context_per_thread *pt = &context->pt[(int)wsi->tsi];
+    struct lws_pollfd eventfd;
+
+    lws_context_lock(pt->context, __func__);
+    lws_pt_lock(pt, __func__);
+
+    if (pt->is_destroyed)
+        goto bail;
+
+    eventfd.fd = fd;
+    eventfd.events = 0;
+    eventfd.revents = 0;
+
+    // TODO handle revents error bits
+
+    if (revents & EPOLLIN) {
+        eventfd.events |= LWS_POLLIN;
+        eventfd.revents |= LWS_POLLIN;
+    }
+
+    if (revents & EPOLLOUT) {
+        eventfd.events |= LWS_POLLOUT;
+        eventfd.revents |= LWS_POLLOUT;
+    }
+
+    lws_pt_unlock(pt);
+    lws_context_unlock(pt->context);
```

Comment:
Is it correct to release the locks here? Because you still use `pt` later.

Reply:
Copied from libuv/libev.

But I will check this again...